ocaml/xapi/xapi_vm_migrate.ml

```diff
@@ -93,31 +93,47 @@ open Storage_interface
 open Listext
 open Fun
 
-let assert_sr_support_migration ~__context ~vdi_map ~remote =
+let assert_sr_support_operations ~__context ~vdi_map ~remote ~ops ~_type =
+  let op_supported_on_source_sr vdi ops =
+    (* Check VDIs must not be present on SR which doesn't have required capability *)
+    let source_sr = Db.VDI.get_SR ~__context ~self:vdi in
+    let sr_record = Db.SR.get_record_internal ~__context ~self:source_sr in
+    let sr_features = Xapi_sr_operations.features_of_sr ~__context sr_record in
+    if not (List.for_all (fun op -> Smint.(has_capability op sr_features)) ops) then
+      raise (Api_errors.Server_error(Api_errors.sr_does_not_support_migration, [Ref.string_of source_sr]));
+  in
+  let op_supported_on_dest_sr sr ops sm_record remote =
+    (* Check VDIs must not be mirrored to SR which doesn't have required capability *)
+    let sr_type = XenAPI.SR.get_type remote.rpc remote.session sr in
+    let sm_capabilities =
+      match List.filter (fun (_, r) -> r.API.sM_type = sr_type) sm_record with
+      | [ _, plugin ] -> plugin.API.sM_capabilities
+      | _ -> []
+    in
+    if not (List.for_all (fun op -> List.mem Smint.(string_of_capability op) sm_capabilities) ops) then
+      raise (Api_errors.Server_error(Api_errors.sr_does_not_support_migration, [Ref.string_of sr]))
+  in
   (* Get destination host SM record *)
   let sm_record = XenAPI.SM.get_all_records remote.rpc remote.session in
-  List.iter (fun (vdi, sr) ->
-      (* Check VDIs must not be present on SR which doesn't have snapshot capability *)
-      let source_sr = Db.VDI.get_SR ~__context ~self:vdi in
-      let sr_record = Db.SR.get_record_internal ~__context ~self:source_sr in
-      let sr_features = Xapi_sr_operations.features_of_sr ~__context sr_record in
-      if not Smint.(has_capability Vdi_snapshot sr_features) then
-        raise (Api_errors.Server_error(Api_errors.sr_does_not_support_migration, [Ref.string_of source_sr]));
-      (* Check VDIs must not be mirrored to SR which doesn't have snapshot capability *)
-      let sr_type = XenAPI.SR.get_type remote.rpc remote.session sr in
-      let sm_capabilities =
-        match List.filter (fun (_, r) -> r.API.sM_type = sr_type) sm_record with
-        | [ _, plugin ] -> plugin.API.sM_capabilities
-        | _ -> []
-      in
-      if not (List.exists (fun cp -> cp = Smint.(string_of_capability Vdi_snapshot)) sm_capabilities) then
-        raise (Api_errors.Server_error(Api_errors.sr_does_not_support_migration, [Ref.string_of sr]))
-    ) vdi_map
+  begin match _type with
+  | `intra_pool ->
+    (* Don't fail if all VDIs going to reside on same SR *)
+    let vdi_map' = List.filter (fun (vdi,sr) -> Db.VDI.get_SR ~__context ~self:vdi <> sr) vdi_map in
+    if vdi_map' <> [] then
```

Comment:
You don't need to test for the empty list. List.iter will simply do nothing when the list is empty. I'd write it like this: `List.filter ... |> List.iter ...` but keeping the `let` is fine, too.